### PR TITLE
Update kotlin generated bytecode to include proper nullability checks and annotations

### PR DESCRIPTION
**Pull request by evanchooly**
There are two parts to this PR.  Functionally, this PR adds back in the nullability checks as done by the kotlinc compiler.  It also brings the generated bytecode in to near identical alignment with the kotlinc generated bytecode.  From the beginning certain aspects of benefits of the kotlin language were lost in the generated bytecode due to missing annotations in jandex with which intellligent choices could be made.  I realized recently, however, that certain assumptions as to nullability can be made about the bytecode because we control and define what that code looks like.  So the generalized solution that would be enabled with proper annotation tracking wasn't strictly necessary in this case.

Given that realization, I've updated the generation to include those nullability checks and annotations on the generated methods to more closely align with how kotlinc does it.  To test this, I added the second part of this PR.  Using the `TraceVisitor` in asm, I can get a list of the bytecode for a given method.  Using this approach, I can simply inspect that list against the "correct" content and be sure out generation is creating the right bytecode.

However, manually constructing each test for each bytecode is not only tedious, but it's error prone and fragile.  It's also incredibly hard to parse visually as a developer trying to debug a failing test.  To correct this, I introduced the notion of "target" method as you'll find in both `Book` and `StudentRepository`.  Know how the generated methods should function (merely passing through the arguments along with the entity type to the Operations class's methods), I can simply manually code all those methods.  Then to validate the generated bytecode, I need only use the `TraceVisitor` mentioned earlier to generate the list of bytecode for the generated and the target method and compare.  If we've done the generation correctly, they'll match.  And this is indeed how it turns out.

*However*, there are a handful of things we can not account for in the bytecode generator.  First among those are labels.  The compiler generates those in ways that asm seems hesitant to allow.  No matter what I tried there were certain locations I simply could not get asm to generate a label.  But largely these labels aren't used.  The one exception is doing the null checks on the return values of certain operations.  But, on the whole, the labels and their names (L0, L1, L2) don't mean much and so the test elides them from the list of bytecodes.  It also updates the `LOCALVARIABLES` table at the end because the start/end label names can vary in way we can't seem to recreate.

And *that* case is typically caused by a `LINENUMBER` bytecode in the middle of the method.  We can't/don't generate this entry because we simply don't know what the line number in the source file is.  As such, the test elides not only the `LINENUMBER` entry but its associated label as well.  This is where the labels get out of sync and why that divergence doesn't really matter.

The last variance that test works around/ignores is the message displayed when a nullability test fails.  This is just the text that the user sees and serves no functional purpose.  The generated code creates a similar, but different, message and so that check would fail without certain tweaks to the bytecode list.  Again this is just cosmetic.

Using this test I was able to quickly update the generated bytecode to match what the compiler does.  This test currently only runs on the MongoDB Panache imperative kotlin code.  However, should this test approach meet wider approval, I would like to run it against the reactive version as well as both the hibernate variants.  Currently this test runs only against kotlinc generated bytecode but there's no reason it could not be run on the javac bytecode as well using the same `target_` approach there.  There might be slightly different "clean ups" to do on the java but should be fairly straightforward to implement those clean ups.  This would help ensure we're generating bytecode on par with what the compiler does.

Why does this matter?  I've had a lingering worry for a while now that certain oddities in our generation might be a problem once we hit native image.  Largely things have seemed to be working just fine but there's enough flakiness on the native image tests that I wanted eliminate as many variables as possible.  But mostly, I trust the bytecode generated by a compiler more than I trust my ability to reason through how it should work.  This test ensures that we hew as close as possible to that example as possible and does it with as little maintenance as possible.

In any case, I am much more confident in the generated bytecode now and going forward than I was when I started.  Thank you for coming to my Ted Talk.

### CODE REVIEW

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/MongoOperations.java [hunk 1]
```diff
@@ -195,6 +196,7 @@ public MongoCollection mongoCollection(Class<?> entityClass) {
         return database.getCollection(entityClass.getSimpleName(), entityClass);
     }
 
+    @NotNull
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/MongoOperations.java:199]: We didn't use this anotation anywhere else so can you give some details of why you add it here ?

File: extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/Book.kt [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import io.quarkus.mongodb.panache.kotlin.PanacheQuery
+import io.quarkus.mongodb.panache.kotlin.runtime.KotlinMongoOperations
+import io.quarkus.mongodb.panache.kotlin.runtime.KotlinMongoOperations.Companion.INSTANCE
+import io.quarkus.panache.common.Parameters
+import io.quarkus.panache.common.Sort
+import org.bson.Document
+import org.bson.types.ObjectId
+import java.util.stream.Stream
+
+@Suppress("UNCHECKED_CAST", "unused")
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/Book.kt:17]: You really need to write some explaination of the content of this class so if someone saw the test fail but didn't know the detail of the bytecode enancement knows what's going on.

File: extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        evaluate(new ByteCodeType(StudentRepository.class), new ByteCodeType(Student.class), PanacheMongoRepositoryBase.class, LONG);
+    }
+
+    /**
+     * we don't generate labels, e.g., like the compiler does (it's sourcemap related) so we need to elide some lines
+     * to account for the difference.
+     *
+     * @return
+     */
+    private List<Object> cleanUp(List<Object> input) {
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java:69]: Can you move this method closer to where it is used ? For ex. after `validate()`.

File: extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                .filter(e -> e.getKey().startsWith("target_"))
+                .forEach(entry -> {
+                    String key = entry.getKey().substring(7);
+                    BytecodeMethod generatedMethod = localMethods.get(key);
+                    BytecodeMethod targetMethod = entry.getValue();
+                    try {
+                        validate(key, generatedMethod.body, targetMethod.body);
+                    } catch (NullPointerException e) {
+                        throw new RuntimeException("could not find " + key);
+                    }
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java:118]: Never try/catch NPE ! OK it's not that important but it's really an anti-pattern.

Comment (loicmathieu) [reply extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java:118]: Anyway, it's a test, so let's just make it fails with an exception

**Review by loicmathieu — COMMENTED** [review]
I'm not expert on bytecode and ASM so you'll need someone else eyes on it but this PR seems a very good improvements and is very cleverly implemented !

I'll send more feedback on a comment.

**Comment by loicmathieu** [comment]
> It also brings the generated bytecode in to near identical alignment with the kotlinc generated bytecode

Great! This is what we should aim each time we generate bytecode.

> Using this approach, I can simply inspect that list against the "correct" content and be sure out generation is creating the right bytecode.

This is awesome, as I understand it you created a "contract" of what should be generated by providing a test class that contains the code that should be generated and check that the code expected is the actual one.
The only caveat is that it would need to be enhanced each time we add a new method ...

> However, should this test approach meet wider approval, I would like to run it against the reactive version as well as both the hibernate variants

OK for me, if we accept it for one variant, all variants should have it.

> Currently this test runs only against kotlinc generated bytecode but there's no reason it could not be run on the javac bytecode as well using the same target_ approach there

This would be great but javac generated bytecode can change from one version to an other. For kotlinc, you fix the version of the pom.xml so there is no risk, but for javac, the test would be run on 8, 11 and the latest (15 today), and when a new version of Java is release you will automatically launch it on the new one. If we want also to run the test on javac (which is a good idea IMHO but will need some investment) maybe we need to run it only on one version (the lower supported for example).

> Thank you for coming to my Ted Talk.

You're welcome ;)

To conclude, I think it's a great idea, and even if it adds maintenance burden and even if the test is not trivial, I think it's a good addition and that other areas of Quarkus (all panache extensions of course but maybe others) can take advantage of. 
I remember for example some PRs to fix how we produce MongoClient beans for multiple database support that would have been easier to implements with such tool.

> But mostly, I trust the bytecode generated by a compiler more than I trust my ability to reason through how it should work

Same for me ;)

I didn't update our bytecode generation often but each time I did it I have a bad feeling that not knowing exactly what it does. I know some people are comfortable with bytecode and ASM but for the other, having such safetynet at the one you just created is a huge improvements, and can even be used in a TDD approach to help creating the bytecode (crafting the Java/Kotlin code in the test class and interating on the bytecode generation part as long as it didn't generates what's expected).

File: extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/Book.kt (hunk 2 above)
Comment (evanchooly) [reply extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/Book.kt:17]: that's a fair point.  i go in to detail in the PR but folks 6 months from now won't know that.  i'll update.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/MongoOperations.java (hunk 1 above)
Comment (evanchooly) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/MongoOperations.java:199]: That looks like an experiment that wasn't cleaned up.  I'll do so now.

File: extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java (hunk 4 above)
Comment (evanchooly) [reply extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java:118]: it does fail with an exception.  this catch just ensures that the missing key (read: method) is logged so the dev knows where to start looking.  Otherwise the error is off in the middle of random code with little context as to the missing method.

But catching NPE is perfectly for reasons as stated here and others.

**Review by FroMage — APPROVED** [review]

File: extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java (hunk 3 above)
Comment (evanchooly) [reply extensions/panache/mongodb-panache-kotlin/deployment/src/test/kotlin/io/quarkus/mongodb/panache/kotlin/deployment/TestEnhancers.java]: done
